test/post_processor/test_dual_from_empirical_frequencies.py

```diff
@@ -47,52 +53,86 @@ def test_empirical_dual(self):
         job = povm_sampler.run([qc], shots=127, povm=measurement)
         pub_result = job.result()[0]
+        self.post_processor = POVMPostProcessor(pub_result)
 
-        observable = SparsePauliOp(
-            ["XI", "YI", num_qubits * "Y", num_qubits * "Z"], coeffs=[1.3, 1.2, -1, 1.4]
-        )
-
-        post_processor = POVMPostProcessor(pub_result)
+    def test_empirical_dual(self):
+        """Test that the method constructs a valid dual."""
+        observable = SparsePauliOp(["XI", "YI", 2 * "Y", 2 * "Z"], coeffs=[1.3, 1.2, -1, 1.4])
 
         with self.subTest("Test canonical dual."):
-            exp_value, std = post_processor.get_expectation_value(observable)
-            self.assertAlmostEqual(exp_value, 1.255719986997053)
-            self.assertAlmostEqual(std, 0.5312929210892221)
+            exp_value, std = self.post_processor.get_expectation_value(observable)
+            self.assertAlmostEqual(exp_value, -1.920301957227794)
+            self.assertAlmostEqual(std, 0.468920055605158)
 
         with self.subTest("Test empirical dual with default arguments."):
-            post_processor.dual = dual_from_empirical_frequencies(
-                povm_post_processor=post_processor
+            self.post_processor.dual = dual_from_empirical_frequencies(
+                povm_post_processor=self.post_processor
             )
-            exp_value, std = post_processor.get_expectation_value(observable)
-            self.assertAlmostEqual(exp_value, 0.9128662937761666)
-            self.assertAlmostEqual(std, 0.40472771240833644)
+            exp_value, std = self.post_processor.get_expectation_value(observable)
+            self.assertAlmostEqual(exp_value, -4.156136105226559)
+            self.assertAlmostEqual(std, 0.10312825405831737)
 
         with self.subTest("Test empirical dual with lists arguments."):
-            post_processor.dual = dual_from_empirical_frequencies(
-                povm_post_processor=post_processor,
+            self.post_processor.dual = dual_from_empirical_frequencies(
+                povm_post_processor=self.post_processor,
                 loc=0,
                 bias=[6, 6],
-                ansatz=[DensityMatrix(np.eye(2) / 2), DensityMatrix(np.eye(2) / 2)],
+                ansatz=[
+                    DensityMatrix(np.eye(2) / 2),
+                    DensityMatrix(np.array([[2, 0], [1, 0]]) / 3),
+                ],
             )
-            exp_value, std = post_processor.get_expectation_value(observable)
-            self.assertAlmostEqual(exp_value, 0.9128662937761666)
-            self.assertAlmostEqual(std, 0.40472771240833644)
+            exp_value, std = self.post_processor.get_expectation_value(observable)
+            self.assertAlmostEqual(exp_value, -4.16623022578294)
+            self.assertAlmostEqual(std, 0.1035227891358374)
 
         with self.subTest(
             "Test empirical dual with `bias` and `ansatz` arguments repeated for all qubits."
```

Comment:
I don't think I understand where this repetition is being tested here? :thinking:

Reply:
It's because we only supply one value of `bias` and `ansatz` (instead of a list specifying a value for each qubit), then this single value is used (or "repeated") for each qubit.

Reply:
Ah I see okay :+1: